File: ChatGPT/clojure/output/output_06b.clj
;; ChatGPT-generated solution will go here.
(ns output_06b
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(def deltas {\^ [-1 0], \> [0 1], \v [1 0], \< [0 -1]})
(def turn-right {\^ \>, \> \v, \v \<, \< \^})

(defn in-bounds? [grid [r c]]
  (and (<= 0 r) (< r (count grid))
       (<= 0 c) (< c (count (first grid)))))

(defn blocked? [grid [r c]]
  (or (not (in-bounds? grid [r c]))
      (= (get-in grid [r c]) \#)))

(defn move [[r c] dir]
  (let [[dr dc] (deltas dir)]
    [(+ r dr) (+ c dc)]))

(defn find-guard [grid]
  (first (for [r (range (count grid))
               c (range (count (first grid)))
               :let [ch (get-in grid [r c])]
               :when (contains? deltas ch)]
           [[r c] ch])))

(defn simulate-path [grid start-pos start-dir]
  (loop [pos start-pos
         dir start-dir
         seen #{}]
    (let [state [pos dir]]
      (cond
        (not (in-bounds? grid pos)) {:loop false}
        (seen state) {:loop true}
        :else
        (let [ahead (move pos dir)]
          (if (blocked? grid ahead)
            (recur pos (turn-right dir) (conj seen state))
            (recur ahead dir (conj seen state))))))))

(defn part [lines]
  (let [grid (mapv vec lines)
        [[start-pos start-dir]] (vector (find-guard grid))
        rows (count grid)
        cols (count (first grid))]
    (count
     (for [r (range rows)
           c (range cols)
           :let [ch (get-in grid [r c])]
           :when (and (not= [r c] start-pos)
                      (not= ch \#))
           :let [new-grid (assoc-in grid [r c] \#)
                 result (simulate-path new-grid start-pos start-dir)]
           :when (:loop result)]
       [r c]))))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))

File: ChatGPT/Direct_Attempt/python/output/output_06b.py
# ChatGPT-generated solution will go here.
def main():
    with open('input.txt') as f:
        grid = [list(line.strip()) for line in f if line.strip()]

    rows, cols = len(grid), len(grid[0])
    directions = ['^', '>', 'v', '<']
    deltas = [(-1, 0), (0, 1), (1, 0), (0, -1)]
    turn_right = {0: 1, 1: 2, 2: 3, 3: 0}

    # Find the guard's starting position and direction
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] in directions:
                start_pos = (r, c)
                dir_idx = directions.index(grid[r][c])
                break

    def simulate(grid, start_pos, dir_idx):
        visited = set()
        seen_states = set()
        pos = start_pos
        visited.add(pos)
        while True:
            state = (pos, dir_idx)
            if state in seen_states:
                return True, visited  # loop detected
            seen_states.add(state)
            dr, dc = deltas[dir_idx]
            nr, nc = pos[0] + dr, pos[1] + dc
            if not (0 <= nr < rows and 0 <= nc < cols):
                return False, visited  # exited grid
            if grid[nr][nc] == '#':
                dir_idx = turn_right[dir_idx]
            else:
                pos = (nr, nc)
                visited.add(pos)

    # Simulate without any modification
    _, original_path = simulate(grid, start_pos, dir_idx)

    # Try placing a new obstruction on every empty cell (except start)
    count = 0
    for r in range(rows):
        for c in range(cols):
            if (r, c) != start_pos and grid[r][c] == '.':
                grid[r][c] = '#'
                loops, _ = simulate(grid, start_pos, dir_idx)
                if loops:
                    count += 1
                grid[r][c] = '.'
    print(count)

if __name__ == '__main__':
    main()
